File: src/transcript.ts
import { homedir } from "node:os";
import { join } from "node:path";
import {
  CAPTURE_BYTE_LIMIT,
  CaptureLimitError,
  TRANSCRIPT_FILE_BYTE_LIMIT,
  TRANSCRIPT_RECORD_BYTE_LIMIT,
  assertUnderCaptureCap,
} from "./caps";
import type { TranscriptExtractor } from "./context";
import type { AgentSessionInfo } from "./host";

const TRANSCRIPT_TIMEOUT_MS = 30_000;

export function slug(cwd: string): string {
  return cwd.replace(/[^a-zA-Z0-9]/g, "-");
}

type ContentBlock = { type?: unknown; text?: unknown };

function extractText(content: unknown): string {
  if (typeof content === "string") return content;
  if (!Array.isArray(content)) return "";
  const parts: string[] = [];
  for (const block of content as ContentBlock[]) {
    if (block?.type === "text" && typeof block.text === "string") parts.push(block.text);
  }
  return parts.join("");
}

function extractEntry(line: string): string | undefined {
  if (!line.trim()) return;
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch {
    return;
  }
  const row = parsed as { type?: unknown; message?: { content?: unknown } };
  if (row.type !== "user" && row.type !== "assistant") return;
  if (!row.message || row.message.content === undefined) return;
  const text = extractText(row.message.content);
  return text ? `${row.type}:\n${text}` : undefined;
}

export function extractAgentTranscript(jsonl: string): string {
  const entries: string[] = [];
  for (const line of jsonl.split("\n")) {
    const entry = extractEntry(line);
    if (entry) entries.push(entry);
  }
  return entries.join("\n\n");
}

function appendTranscriptEntry(entries: string[], entry: string, bytes: number): number {
  const nextBytes = bytes + (entries.length ? 2 : 0) + Buffer.byteLength(entry);
  if (nextBytes > CAPTURE_BYTE_LIMIT) {
    throw new CaptureLimitError("transcript", nextBytes);
  }
  entries.push(entry);
  return nextBytes;
}

async function readClaudeTranscriptStream(file: Bun.BunFile): Promise<string> {
  const decoder = new TextDecoder();
  let line = "";
  const entries: string[] = [];
  let transcriptBytes = 0;
  let bytesRead = 0;

  const consumeLine = (completedLine: string): void => {
    const recordBytes = Buffer.byteLength(completedLine);
    if (recordBytes > TRANSCRIPT_RECORD_BYTE_LIMIT) {
      throw new CaptureLimitError("transcript record", recordBytes, TRANSCRIPT_RECORD_BYTE_LIMIT);
    }
    const entry = extractEntry(completedLine);
    if (entry) transcriptBytes = appendTranscriptEntry(entries, entry, transcriptBytes);
  };

  for await (const chunk of file.stream()) {
    bytesRead += chunk.byteLength;
    if (bytesRead > TRANSCRIPT_FILE_BYTE_LIMIT) {
      throw new CaptureLimitError("transcript file", bytesRead, TRANSCRIPT_FILE_BYTE_LIMIT);
    }
    line += decoder.decode(chunk, { stream: true });
    let newline;
    while ((newline = line.indexOf("\n")) !== -1) {
      consumeLine(line.slice(0, newline));
      line = line.slice(newline + 1);
    }
    // UTF-16 length lower-bounds UTF-8 bytes: the partial record is rejected
    // before it grows unbounded without an O(n^2) byte count per chunk.
    if (line.length > TRANSCRIPT_RECORD_BYTE_LIMIT) {
      throw new CaptureLimitError(
        "transcript record",
        Buffer.byteLength(line),
        TRANSCRIPT_RECORD_BYTE_LIMIT,
      );
    }
  }
  line += decoder.decode();
  consumeLine(line);
  return entries.join("\n\n");
}

export async function readClaudeTranscript(
  cwd: string,
  sessionId: string,
  base = join(homedir(), ".claude", "projects"),
): Promise<string> {
  const { HerdrError } = await import("./host");
  const path = join(base, slug(cwd), `${sessionId}.jsonl`);
  const file = Bun.file(path);
  if (!(await file.exists())) {
    throw new HerdrError("transcript_file_missing", `transcript file not found: ${path}`);
  }
  try {
    const size = file.size;
    if (size > TRANSCRIPT_FILE_BYTE_LIMIT) {
      throw new CaptureLimitError("transcript file", size, TRANSCRIPT_FILE_BYTE_LIMIT);
    }
    return await readClaudeTranscriptStream(file);
  } catch (error) {
    if (error instanceof HerdrError || error instanceof CaptureLimitError) throw error;
    throw new HerdrError(
      "transcript_file_unreadable",
      `transcript file unreadable: ${path}${error instanceof Error ? ` (${error.message})` : ""}`,
    );
  }
}

function transcriptEnv(
  paneId: string,
  info: AgentSessionInfo,
  invocationCwd: string,
): NodeJS.ProcessEnv {
  const cwd = info.cwd || invocationCwd;
  const env: NodeJS.ProcessEnv = {
    ...process.env,
    HWF_TRANSCRIPT_PANE_ID: paneId,
    HWF_TRANSCRIPT_AGENT_KIND: info.agent,
    HWF_TRANSCRIPT_CWD: cwd,
  };
  if (info.sessionKind) env.HWF_TRANSCRIPT_SESSION_KIND = info.sessionKind;
  if (info.sessionId) env.HWF_TRANSCRIPT_SESSION_VALUE = info.sessionId;
  return env;
}

async function runTranscriptCommand(
  argv: string[],
  paneId: string,
  info: AgentSessionInfo,
  invocationCwd: string,
): Promise<string> {
  const { HerdrError } = await import("./host");
  const cwd = info.cwd || invocationCwd;
  const { spawnCapture } = await import("./engine");
  const result = await spawnCapture(argv, {
    cwd,
    env: transcriptEnv(paneId, info, invocationCwd),
    timeoutMs: TRANSCRIPT_TIMEOUT_MS,
    maxCaptureBytes: { source: "transcript" },
  });

  if (result.timedOut) {
    throw new HerdrError(
      "transcript_command_failed",
      `transcript command for '${info.agent}' failed: timed out after ${result.timeoutMs / 1000}s`,
    );
  }
  if (result.exitCode !== 0) {
    const tail = result.stderr.trim().slice(-500) || `exit ${result.exitCode}`;
    throw new HerdrError(
      "transcript_command_failed",
      `transcript command for '${info.agent}' failed: ${tail}`,
    );
  }
  if (!result.stdout.trim()) {
    throw new HerdrError(
      "transcript_command_empty",
      `transcript command for '${info.agent}' printed nothing`,
    );
  }
  assertUnderCaptureCap("transcript", result.stdout);
  return result.stdout;
}

export function hasTranscriptSupport(
  agentKind: string,
  transcripts: Record<string, TranscriptExtractor>,
): boolean {
  return agentKind in transcripts || agentKind === "claude";
}

export async function transcriptText(
  paneId: string,
  transcripts: Record<string, TranscriptExtractor> = {},
  opts: {
    invocationCwd: string;
    projectsBase?: string;
    getInfo?: (paneId: string) => Promise<AgentSessionInfo>;
  },
): Promise<string> {
  const { HerdrError, agentSessionInfo } = await import("./host");
  const getInfo = opts.getInfo ?? agentSessionInfo;
  const info = await getInfo(paneId);
  const extractor = transcripts[info.agent];
  if (extractor) {
    return runTranscriptCommand(extractor.command, paneId, info, opts.invocationCwd);
  }
  if (!hasTranscriptSupport(info.agent, transcripts)) {
    throw new HerdrError(
      "transcript_unsupported_kind",
      `no transcript extractor for '${info.agent}' and no built-in support for that kind`,
    );
  }
  const cwd = info.cwd || opts.invocationCwd;
  if (!info.sessionId) {
    throw new HerdrError(
      "transcript_unsupported_kind",
      `no transcript extractor for '${info.agent}' and built-in support requires a native session value`,
    );
  }
  return readClaudeTranscript(cwd, info.sessionId, opts.projectsBase);
}

File: internal/transcript/transcript.go
// Package transcript reads agent transcripts. It has a built-in Claude
// .jsonl extractor and, for other agent kinds, runs a configured extractor
// command.
package transcript

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/aorumbayev/herdr-workflows/internal/caps"
	"github.com/aorumbayev/herdr-workflows/internal/host"
)

var nonAlnumRE = regexp.MustCompile(`[^a-zA-Z0-9]`)

// Slug must reproduce the project directory name Claude Code writes, which
// comes from a JavaScript `String.replace` over UTF-16 code units: an astral
// rune is two units and becomes two dashes.
func Slug(cwd string) string {
	return nonAlnumRE.ReplaceAllStringFunc(cwd, func(match string) string {
		if r := []rune(match)[0]; r > 0xFFFF {
			return "--"
		}
		return "-"
	})
}

func extractText(content any) string {
	switch c := content.(type) {
	case string:
		return c
	case []any:
		var parts []string
		for _, block := range c {
			m, ok := block.(map[string]any)
			if !ok {
				continue
			}
			if m["type"] == "text" {
				if text, ok := m["text"].(string); ok {
					parts = append(parts, text)
				}
			}
		}
		return strings.Join(parts, "")
	default:
		return ""
	}
}

func extractEntry(line []byte) string {
	if strings.TrimSpace(string(line)) == "" {
		return ""
	}
	var row struct {
		Type    string `json:"type"`
		Message *struct {
			Content any `json:"content"`
		} `json:"message"`
	}
	if err := json.Unmarshal(line, &row); err != nil {
		return ""
	}
	if row.Type != "user" && row.Type != "assistant" {
		return ""
	}
	if row.Message == nil || row.Message.Content == nil {
		return ""
	}
	text := extractText(row.Message.Content)
	if text == "" {
		return ""
	}
	return row.Type + ":\n" + text
}

func appendEntry(entries *[]string, transcriptBytes *int, entry string) error {
	next := *transcriptBytes + len(entry)
	if len(*entries) > 0 {
		next += 2
	}
	if next > caps.CaptureByteLimit {
		return &caps.CaptureLimitError{Source: "transcript", Bytes: next, Limit: caps.CaptureByteLimit}
	}
	*entries = append(*entries, entry)
	*transcriptBytes = next
	return nil
}

func consumeLine(line []byte, entries *[]string, transcriptBytes *int) error {
	if len(line) > caps.TranscriptRecordByteLimit {
		return &caps.CaptureLimitError{Source: "transcript record", Bytes: len(line), Limit: caps.TranscriptRecordByteLimit}
	}
	if entry := extractEntry(line); entry != "" {
		return appendEntry(entries, transcriptBytes, entry)
	}
	return nil
}

// consumeChunk drains every complete line out of pending and returns the
// unterminated remainder.
func consumeChunk(pending []byte, entries *[]string, transcriptBytes *int) ([]byte, error) {
	for {
		i := bytes.IndexByte(pending, '\n')
		if i == -1 {
			break
		}
		if err := consumeLine(pending[:i], entries, transcriptBytes); err != nil {
			return nil, err
		}
		pending = pending[i+1:]
	}
	if len(pending) > caps.TranscriptRecordByteLimit {
		return nil, &caps.CaptureLimitError{Source: "transcript record", Bytes: len(pending), Limit: caps.TranscriptRecordByteLimit}
	}
	return pending, nil
}

func readStream(r io.Reader) (string, error) {
	var entries []string
	transcriptBytes := 0
	var pending []byte
	bytesRead := 0
	buf := make([]byte, 64*1024)
	for {
		n, readErr := r.Read(buf)
		if n > 0 {
			bytesRead += n
			if bytesRead > caps.TranscriptFileByteLimit {
				return "", &caps.CaptureLimitError{Source: "transcript file", Bytes: bytesRead, Limit: caps.TranscriptFileByteLimit}
			}
			var err error
			if pending, err = consumeChunk(append(pending, buf[:n]...), &entries, &transcriptBytes); err != nil {
				return "", err
			}
		}
		if readErr != nil {
			if errors.Is(readErr, io.EOF) {
				break
			}
			return "", readErr
		}
	}
	if len(pending) > 0 {
		if err := consumeLine(pending, &entries, &transcriptBytes); err != nil {
			return "", err
		}
	}
	return strings.Join(entries, "\n\n"), nil
}

// ReadClaudeTranscript loads a Claude session .jsonl and returns the extracted
// user/assistant text. base is the projects directory the session lives under.
func ReadClaudeTranscript(cwd, sessionID, base string) (string, error) {
	path := filepath.Join(base, Slug(cwd), sessionID+".jsonl")
	f, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return "", &host.HerdrError{Code: "transcript_file_missing", Msg: "transcript file not found: " + path}
		}
		return "", &host.HerdrError{Code: "transcript_file_unreadable", Msg: unreadableMsg(path, err)}
	}
	defer func() { _ = f.Close() }()
	info, err := f.Stat()
	if err != nil {
		return "", &host.HerdrError{Code: "transcript_file_unreadable", Msg: unreadableMsg(path, err)}
	}
	if info.Size() > caps.TranscriptFileByteLimit {
		return "", &caps.CaptureLimitError{Source: "transcript file", Bytes: int(info.Size()), Limit: caps.TranscriptFileByteLimit}
	}
	text, err := readStream(f)
	if err != nil {
		var capErr *caps.CaptureLimitError
		if errors.As(err, &capErr) {
			return "", err
		}
		return "", &host.HerdrError{Code: "transcript_file_unreadable", Msg: unreadableMsg(path, err)}
	}
	return text, nil
}

func unreadableMsg(path string, err error) string {
	if err != nil {
		return "transcript file unreadable: " + path + " (" + err.Error() + ")"
	}
	return "transcript file unreadable: " + path
}
